### PR TITLE
Inline defaulted functions

### PR DESCRIPTION
**Pull request by Thomas1664**
In the case of out of line defaulted move constructors this prevented some types from being nothrow movable.
Please also note that associative containers aren't nothrow movable on MSVC. One has to explicitly mark default move constructors as `noexcept` for types containing associative containers to make them nothrow movable.

### CODE REVIEW

File: include/vcpkg/platform-expression.h [hunk 1]
```diff
@@ -25,14 +25,14 @@ namespace vcpkg::PlatformExpression
         static Expr Or(std::vector<Expr>&& exprs);
 
         // since ExprImpl is not yet defined, we need to define the ctor and dtor in the C++ file
-        Expr(); // always true
+        Expr() noexcept; // always true
```
Comment (BillyONeal) [inline include/vcpkg/platform-expression.h:28]: I disagree with this change. These aren't nothrow default constructible or nothrow movable and I see no reason to insert code to enforce termination here.

File: include/vcpkg/platform-expression.h [hunk 2]
```diff
@@ -25,14 +25,14 @@ namespace vcpkg::PlatformExpression
...
         Expr(const Expr&);
-        Expr(Expr&&);
+        Expr(Expr&&) noexcept;
         Expr& operator=(const Expr& e);
-        Expr& operator=(Expr&&);
+        Expr& operator=(Expr&&) noexcept;
 
         explicit Expr(std::unique_ptr<detail::ExprImpl>&& e);
-        ~Expr();
+        ~Expr() noexcept;
```
Comment (BillyONeal) [inline include/vcpkg/platform-expression.h:35]:
```suggestion
        ~Expr();
```

Destructors default to noexcept.

File: include/vcpkg/vcpkgpaths.h [hunk 3]
```diff
@@ -50,7 +50,7 @@ namespace vcpkg
         VcpkgPaths(const Filesystem& filesystem, const VcpkgCmdArguments& args, const BundleSettings& bundle);
         VcpkgPaths(const VcpkgPaths&) = delete;
         VcpkgPaths& operator=(const VcpkgPaths&) = delete;
-        ~VcpkgPaths();
+        ~VcpkgPaths() noexcept;
```
Comment (BillyONeal) [inline include/vcpkg/vcpkgpaths.h:53]: Ditto destructors shouldn't be touched.

File: include/vcpkg/xunitwriter.h [hunk 4]
```diff
@@ -21,7 +21,7 @@ namespace vcpkg
     public:
         // Out of line con/destructor avoids exposing XunitTest
         XunitWriter();
-        ~XunitWriter();
+        ~XunitWriter() noexcept;
```
Comment (BillyONeal) [inline include/vcpkg/xunitwriter.h:24]: Ditto dtor

**Review by BillyONeal — CHANGES_REQUESTED** [review]
This is reverting an explicit design choice to favor build throughput by not needing the definition of these functions spammed out into every object file in places where performance does not matter.

Do you have data showing this is meaningful for customers?

File: include/vcpkg/platform-expression.h (hunk 1 above)
Comment (Thomas1664) [reply include/vcpkg/platform-expression.h:28]: Why is Expr not nothrow (move) constructible?

**Comment by Thomas1664** [comment]
> This is reverting an explicit design choice to favor build throughput by not needing the definition of these functions spammed out into every object file in places where performance does not matter.
> 
> Do you have data showing this is meaningful for customers?

The main goal of this PR is to avoid not marking move constructors as noexcept in the case of out of line defaulted move constructors. I highly doupt that inlining vs not inlining makes a difference at compile time or run time.

File: include/vcpkg/platform-expression.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/platform-expression.h:28]: I think I thought there was a `std::map` in here.

**Comment by BillyONeal** [comment]
@ras0219-msft @data-queue @vicroms @JavierMatosD and I discussed this today and we are happy with this so long as defaulting in the header does not require exposing additional dependencies in headers (e.g. in pimpl patterns like this:

```c++
// header.hpp

#include <memory>

struct impl;

struct ptr {
    std::unique_ptr<impl> instance;
    // ~ptr() = default; // error: impl is incomplete, how do I ~impl?
    ~ptr();
};
```

)

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks :)

**Pull request merged**